carbon/job.py

```diff
@@ -75,40 +84,113 @@ class Job:
     cputime: float
     """The total CPU time used by the job in core-hours."""
 
-    ngpus: int
-    """The number of GPUs used by the job."""
+    gputime: float
+    """The total GPU time used by the job in component-hours."""
 
-    memory: float
-    """The memory allocated to the job in GB."""
+    memtime: float
+    """The total memory-time allocated to the job in GB-hours."""
 
     node: str
     """The node the job was executed on."""
 
+    state: JobState = JobState.FINISHED
+    """The state of the job."""
+
     @classmethod
-    def fromPBS(cls, id: str) -> Self:
-        """Create a Job object by fetching data from PBS based on the job ID.
+    def is_array(cls, job_id: str) -> bool:
+        """Is it a PBS array job?"""
+        return job_id.split(".")[0].endswith("[]")
+
+    @classmethod
+    def split_sub_jobs(cls, job_id: str) -> list[str]:
+        """Split a PBS array job id into the corresponding list of subjob ids."""
+        # Job ID for array should be digits followed by square brackets
+        if not re.fullmatch(r"^\d+\[\](?:..*)?$", job_id):
+            raise MalformedJobIDError(
+                f"Malformed array job ID: {job_id}. Should contain only digits, "
+                "followed by square brackets"
+            )
+
+        cmd = f"qstat -xJt {job_id}"
+
+        try:
+            output = subprocess.run(
+                cmd,
+                shell=True,
+                check=True,
+                timeout=20,
+                capture_output=True,
+                text=True,  # Potentially a 10000 line long str. Could improve?
+            )
+        except subprocess.CalledProcessError as e:
+            if e.returncode == 153:
+                raise UnknownJobIDError(f"Unknown job ID: {job_id}")
+            elif e.returncode == 1 or e.returncode == 170:
+                raise MalformedJobIDError(f"Malformed job ID: {job_id}")
+            else:
+                raise ValueError(f"Failed to fetch job data: {e}")
+
+        sub_jobs = []
+        for row in output.stdout.splitlines()[3:]:
+            items = row.split()
+            label = items[0]
+            state = items[4]
+            # Get all the subjobs which are running, finished, or expired (finished but
+            # other subjobs are still running).
+            if re.fullmatch(r"^\d+\[\d+\](?:..*)?$", label) and state in [
```

Comment:
As this may be run many times consider using `re.compile` before the loop to reduce overhead.